dask/layers.py

```diff
@@ -1239,3 +1262,341 @@ def __repr__(self):
         return "DataFrameIOLayer<name='{}', n_parts={}, columns={}>".format(
             self.name, len(self.inputs), self.columns
         )
+
+
+class DataFrameTreeReduction(DataFrameLayer):
+    """DataFrame Tree-Reduction Layer
+
+    Parameters
+    ----------
+    name : str
+        Name to use for the constructed layer.
+    name_input : str
+        Name of the input layer that is being reduced.
+    npartitions_input : str
+        Number of partitions in the input layer.
+    concat_func : callable
+        Function used by each tree node to reduce a list inputs into
+        a single output value. This function must accept only a list
+        as its first positional argument.
+    tree_node_func : callable
+        Function used on the output of ``concat_func`` in each tree
+        node. This function must accept the output of ``concat_func``
+        as its first positional argument. Any other input arguments
+        must be definded by in ``tree_node_kwargs``.
+    tree_node_args : dict, optional
+        Dictionary of key-word arguments to include in every call to
+        the function specified by ``tree_node_func``. Note that the
+        first positional argument will always correspond to a list of
+        inputs from the input nodex (and these arguments will always
+        come after).
+    finalize_func : callable, optional
+        Function used in place of ``tree_node_func`` on the final tree
+        node(s) to produce the final output for each split. By default,
+        ``tree_node_func`` will be used.
+    finalize_kwargs : dict, optional
+        Dictionary of key-word arguments to include in every call to
+        the function specified by ``finalize_func``. This parameter
+        is ignored if ``finalize_func`` is not set.
+    split_every : int, optional
+        This argument specifies the maximum number of input nodes
+        inputs to be handled by any one task in the tree. Defaults to 32.
+    split_out : int, optional
+        This argument specifies the number of output nodes in the
+        reduction tree. If ``split_out`` is set to an integer >=1, the
+        input tasks must contain data that can be indexed by a ``getitem``
+        operation with a key in the range ``[0, split_out)``.
+    output_splits : list, optional
```

Comment:
I'm confused by this name. What about `output_partitions`? `output_indices`? `outputs`?

Reply:
`output_partitions` makes sense to me.